testng-core/src/main/java/org/testng/internal/invokers/MethodInvocationHelper.java

```diff
@@ -344,9 +351,12 @@ private static void invokeWithTimeoutWithNoExecutor(
         }
         testResult.setThrowable(e);
       } else {
-        testResult.setThrowable(new ThreadTimeoutException(tm, realTimeOut));
+        if (!(ex instanceof TestNotInvokedException)) {
```

Comment:
> Another option is catching TestNotInvokedException and doing the expected work for this specific exception.

That wont be possible because we are doing something if the exception type is NOT `TestNotInvokedException`

Reply:
> I'm not feeling it good to have a specific workflow for TestNotInvokedException.

We are not having any specific workflow here. We are classifying a problem as timedout only if the problem was NOT of type `TestNotInvokedException` which is meant to hint at the fact that the callback functionality was NOT dealt with in a proper way.

Reply:
What about:

```
} catch (TestNotInvokedException tne) {
      testResult.setThrowable(tne);
      testResult.setStatus(ITestResult.FAILURE);
      return wasInvoked;
} catch (Exception ex) {
  ...
}
```

Reply:
We are not going to gain anything by doing this because now we would be having multiple return statements and we are going to be doing the same thing. I don't see the benefit of this optimisation, unless I am missing something here.

Reply:
We gain in complexity. It is an uncommon pattern the check the exception type in the catch block.
Instead, each type is supposed to be managed in its own block.
If many exceptions have the same logic, they can share the same block.
If many exceptions share some logic, they can use the same method.